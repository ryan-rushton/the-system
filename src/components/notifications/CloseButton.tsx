import { motion } from 'framer-motion';
import { FC } from 'react';
import styles from './CloseButton.module.scss';

const Path = ({ d }: { d: string }) => (
  <motion.path fill="transparent" strokeWidth="3" stroke="#fefefe" strokeLinecap="round" d={d} />
);

interface Props {
  /** A function to be called when the close icon is clicked. */
  onClose(): void;
}

/**
 * A button that creates a close icon for the notifications. Extracted to it's own component so we can
 * isolate the svg creation.
 */
const CloseButton: FC<Props> = ({ onClose }) => (
  <button data-testid="close-button" onClick={onClose} className={styles.close}>
    <svg width="12" height="12" viewBox="0 0 12 12">
      <Path d="M 1 1 L 11 11" />
      <Path d="M 1 11 L 11 1" />
    </svg>
  </button>
);

export default CloseButton;
