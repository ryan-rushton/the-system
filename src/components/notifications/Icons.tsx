import {
  faCheckCircle,
  faExclamationCircle,
  faExclamationTriangle,
  faInfoCircle,
} from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { FC } from 'react';
import styles from './Icons.module.scss';

/** Error icon for notifications. */
export const Error: FC = () => (
  <div data-testid="error-icon" className={styles.icon}>
    <FontAwesomeIcon icon={faExclamationCircle} size="lg" />
  </div>
);

/** Warning icon for notifications. */
export const Warning: FC = () => (
  <div data-testid="warning-icon" className={styles.icon}>
    <FontAwesomeIcon icon={faExclamationTriangle} size="lg" />
  </div>
);

/** Info icon for notifications. */
export const Info: FC = () => (
  <div data-testid="info-icon" className={styles.icon}>
    <FontAwesomeIcon icon={faInfoCircle} size="lg" />
  </div>
);

/** Success icon for notifications. */
export const Success: FC = () => (
  <div data-testid="success-icon" className={styles.icon}>
    <FontAwesomeIcon icon={faCheckCircle} size="lg" />
  </div>
);
