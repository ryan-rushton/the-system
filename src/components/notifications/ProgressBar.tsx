import { motion } from 'framer-motion';
import { FC } from 'react';
import styles from './ProgressBar.module.scss';

interface Props {
  /** The time in seconds at which the animation will finish and onCompletion will be invoked. */
  duration?: number;
  /** A function that gets called after the duration has completed. */
  onCompletion?(): void;
}

/**
 * A animated bar to indicate that the notification will close itself after the nominated duration.
 */
const ProgressBar: FC<Props> = ({ duration, onCompletion }) => {
  if (duration === undefined) {
    return null;
  }

  return (
    <div data-testid="progress-bar" className={styles.container}>
      <motion.div
        className={styles.progress}
        initial={{ width: '0%' }}
        animate={{ width: '100%', transition: { duration } }}
        onAnimationComplete={onCompletion}
      />
    </div>
  );
};

export default ProgressBar;
