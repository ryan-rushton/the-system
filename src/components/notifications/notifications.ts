import { Subject } from 'rxjs';

// An id for each notification so we know which ones to clear.
let id = 0;

export interface Notification {
  /** The severity of the notification, this will determine the colour displayed. */
  severity: 'info' | 'warning' | 'error' | 'success';
  /** The notification message to show the user. */
  message: string;
  /** Where the notification originated from, allows us to conditionally clear old notifications. */
  origin: string;
  /** How long the notification is displayed for. If undefined the notification stays until closed. */
  duration?: number;
}

export interface UniqueNotification extends Notification {
  /** A unique identifier for the notification. */
  id: number;
}

/** An rxjs Subject for notifications. */
export const notifications$ = new Subject<UniqueNotification>();

/** Queues a new notification to be shown in the notification view. */
export const showNotification = (nextNotification: Notification): void => {
  notifications$.next({ ...nextNotification, id: id++ });
};
