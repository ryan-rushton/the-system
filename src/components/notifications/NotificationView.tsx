import clsx from 'clsx';
import { motion } from 'framer-motion';
import { CloseButton } from './CloseButton';
import { ErrorIcon, InfoIcon, SuccessIcon, WarningIcon } from './Icons';
import styles from './NotificationView.module.scss';
import { ProgressBar } from './ProgressBar';
import { UniqueNotification } from './notifications';

/**
 * A view for a single  This includes the close button and the progress bar.
 * When the progress bar fills the notification will close.
 */
export function NotificationView({
  notification: { severity, message, duration },
  onClose,
}: {
  /** The notification to show. */
  notification: UniqueNotification;
  /** Function to call when the close button is pressed or the duration runs out. */
  onClose(): void;
}) {
  return (
    <motion.li
      data-testid="notification-view"
      className={clsx(styles.notification, {
        [styles.success]: severity === 'success',
        [styles.info]: severity === 'info',
        [styles.warning]: severity === 'warning',
        [styles.error]: severity === 'error',
      })}
      initial={{ opacity: 0, scale: 0.3, y: 50 }}
      animate={{ opacity: 1, scale: 1, y: 0 }}
      exit={{
        opacity: 0,
        scale: 0.5,
        transition: { duration: 0.3 },
      }}
    >
      <div className={styles.content}>
        {severity === 'error' && <ErrorIcon />}
        {severity === 'warning' && <WarningIcon />}
        {severity === 'info' && <InfoIcon />}
        {severity === 'success' && <SuccessIcon />}
        <div className={styles.message}>{message}</div>
        <CloseButton onClose={onClose} />
      </div>
      {duration !== undefined && <ProgressBar duration={duration} onCompletion={onClose} />}
    </motion.li>
  );
}
