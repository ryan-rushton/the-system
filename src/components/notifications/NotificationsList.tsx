import { AnimatePresence } from 'framer-motion';
import React, { FC, useEffect, useState } from 'react';
import NotificationView from './NotificationView';
import styles from './NotificationsList.module.scss';
import { UniqueNotification, notifications$ } from './notifications';

/**
 * A view for notifications. This should be placed on either side of the screen. It will take up a 20vw column that
 * is nearly the height of the screen.
 *
 * Notifications will have a close icon and follower notifications will disappear when a new follower notification
 * is shown.
 */
const NotificationsList: FC = () => {
  const [notifications, setNotifications] = useState<UniqueNotification[]>([]);

  // subscribe to the notifications stream and update state when new ones come in.
  useEffect(() => {
    const subscription = notifications$.subscribe((newNotification: UniqueNotification): void => {
      setNotifications((oldState) => {
        const newState =
          newNotification.origin === 'follower' ? oldState.filter((n) => n.origin !== 'follower') : [...oldState];
        return [...newState, newNotification];
      });
    });

    return (): void => subscription.unsubscribe();
  }, []);

  return (
    <div data-testid="notifications-list" className={styles.container}>
      <ul className={styles.messageList}>
        <AnimatePresence initial={false}>
          {notifications.map((notification) => (
            <NotificationView
              key={notification.id}
              notification={notification}
              onClose={() => {
                setNotifications((oldState) => oldState.filter((n) => n.id !== notification.id));
              }}
            />
          ))}
        </AnimatePresence>
      </ul>
    </div>
  );
};

export default NotificationsList;
