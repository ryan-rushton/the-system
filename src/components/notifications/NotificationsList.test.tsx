import { act, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, test } from 'vitest';
import { NotificationsList } from './NotificationsList';
import { showNotification } from './notifications';

describe('NotificationsList', () => {
  test('it renders', () => {
    render(<NotificationsList />);

    expect(screen.getByTestId('notifications-list')).toBeInTheDocument();
  });

  test('it matches snapshot', () => {
    render(<NotificationsList />);

    expect(screen.getByTestId('notifications-list')).toMatchSnapshot();
  });

  test('it renders a notification', async () => {
    render(<NotificationsList />);
    act(() => showNotification({ origin: 'origin', severity: 'error', message: 'Some error!' }));
    await screen.findByTestId('notification-view');
    expect(screen.getByText(/^Some error!$/)).toBeInTheDocument();
  });

  test('can close notification', async () => {
    render(<NotificationsList />);

    // show notification and check it's present
    act(() => showNotification({ origin: 'origin', severity: 'error', message: 'Some error!' }));
    await screen.findByTestId('notification-view');
    expect(screen.getByText(/^Some error!$/)).toBeInTheDocument();

    // click the close button and check it goes away
    await userEvent.click(screen.getByTestId('close-button'));
    await waitFor(() => expect(screen.queryByText(/^Some error!$/)).not.toBeInTheDocument());
  });
});
