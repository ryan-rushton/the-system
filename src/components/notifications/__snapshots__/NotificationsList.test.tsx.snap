// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NotificationsList > it matches snapshot 1`] = `
<div
  class="_container_9abc2b"
  data-testid="notifications-list"
>
  <ul
    class="_messageList_9abc2b"
  />
</div>
`;
